app/frontend/wallet/cardano-wallet.js

```diff
@@ -162,29 +162,37 @@ const CardanoWallet = async (options) => {
     return await blockchainExplorer.fetchTxInfo(txHash)
   }
 
+  function dropUnprofitableUtxos(utxos) {
+    const profitableUtxos = []
+    const witnessSize = 139
+    let addedCost = 0
+    let inputSize = 0
+    for (const utxo of utxos) {
+      inputSize = cbor.encode(TxInputFromUtxo(utxo)).length
+      //txFeeFunction(0) returns just the static part of fee that needs to be paid anyway
+      addedCost = txFeeFunction(inputSize + witnessSize) - txFeeFunction(0)
+      if (utxo.coins > addedCost) {
+        //only if a given unspent output has more coins than what we would spend on fees to add
+        //that unspent output, we keep it
+        profitableUtxos.push(utxo)
+      }
+    }
+    return profitableUtxos
+  }
+
   async function prepareTxInputs(coins, address) {
     // we do it pseudorandomly to guarantee fee computation stability
     const randomGenerator = PseudoRandom(state.randomSeed)
     const utxos = shuffleArray(await getUnspentTxOutputs(), randomGenerator)
+    const profitableUtxos = dropUnprofitableUtxos(utxos)
```

Comment:
I suggest making a function for evaluating if a single utxo is profitable and ussing array filter there. Something like this:
```
function isUtxoProfitable(utxo) {
    const inputSize = cbor.encode(TxInputFromUtxo(utxo)).length
    const addedCost = txFeeFunction(inputSize + witnessSize) - txFeeFunction(0)

    return utxo.coins > addedCost
}
...
const profitableUtxos = utxos.filter(isUtxoProfitable)
```